lib/auth/recordingencryption/manager.go

```diff
@@ -0,0 +1,386 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package recordingencryption
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"iter"
+	"log/slog"
+	"slices"
+
+	"filippo.io/age"
+	"github.com/gravitational/trace"
+
+	recordingencryptionv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/recordingencryption/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// EncryptionKeyStore provides methods for interacting with encryption keys.
+type EncryptionKeyStore interface {
+	NewEncryptionKeyPair(ctx context.Context, purpose cryptosuites.KeyPurpose) (*types.EncryptionKeyPair, error)
+	GetDecrypter(ctx context.Context, keyPair *types.EncryptionKeyPair) (crypto.Decrypter, error)
+}
+
+// ManagerConfig captures all of the dependencies required to instantiate a Manager.
+type ManagerConfig struct {
+	Backend  services.RecordingEncryption
+	KeyStore EncryptionKeyStore
+	Logger   *slog.Logger
+}
+
+// NewManager returns a new Manager using the given ManagerConfig.
+func NewManager(cfg ManagerConfig) (*Manager, error) {
+	if cfg.Logger == nil {
+		cfg.Logger = slog.New(slog.DiscardHandler)
+	}
+
+	if cfg.Backend == nil {
+		return nil, trace.BadParameter("backend is required")
+	}
+
+	if cfg.KeyStore == nil {
+		return nil, trace.BadParameter("key store is required")
+	}
+
+	return &Manager{
+		RecordingEncryption: cfg.Backend,
+		keyStore:            cfg.KeyStore,
+		logger:              cfg.Logger,
+	}, nil
+}
+
+// A Manager wraps a services.RecordingEncryption and EncryptionKeyStore in order to provide more complex operations
+// than the CRUD methods exposed by services.RecordingEncryption. It primarily handles resolving RecordingEncryption
+// state and searching for accessible decryption keys.
+type Manager struct {
+	services.RecordingEncryption
+
+	logger   *slog.Logger
+	keyStore EncryptionKeyStore
+	uploader events.MultipartUploader
+}
+
+// ResolveRecordingEncryption examines the current state of the RescordingEncryption resource and advances it to the
+// next state on behalf of the current auth server. At a high level it will provision unfulfilled keys for itself,
+// fulfill keys for keystore configurations when possible, and move its own keys through rotation states.
+func (m *Manager) ResolveRecordingEncryption(ctx context.Context) (*recordingencryptionv1.RecordingEncryption, error) {
```

Comment:
Do you think we can split this function into smaller ones?
It feels like it's doing a lot of things, and it has multiple exits for the happy path, depending on the state.

Reply:
Absolutely! I felt the same way but I've been working on resisting my urge break things down too soon 😅 I think this function definitely deserves to be broken down, though

Reply:
I broke this up a bit and removed duplicate decryptions of the same recording encryption key when fulfilling other keys.